keel-core/src/test/kotlin/com/netflix/spinnaker/keel/validators/DeliveryConfigValidatorTests.kt

```diff
@@ -0,0 +1,164 @@
+package com.netflix.spinnaker.keel.validators
+
+import com.netflix.spinnaker.keel.api.artifacts.DockerArtifact
+import com.netflix.spinnaker.keel.core.api.DependsOnConstraint
+import com.netflix.spinnaker.keel.core.api.SubmittedDeliveryConfig
+import com.netflix.spinnaker.keel.core.api.SubmittedEnvironment
+import com.netflix.spinnaker.keel.core.api.SubmittedResource
+import com.netflix.spinnaker.keel.exceptions.DuplicateArtifactReferenceException
+import com.netflix.spinnaker.keel.exceptions.DuplicateResourceIdException
+import com.netflix.spinnaker.keel.exceptions.MissingEnvironmentReferenceException
+import com.netflix.spinnaker.keel.test.DummyResourceSpec
+import com.netflix.spinnaker.keel.test.TEST_API_V1
+import dev.minutest.junit.JUnit5Minutests
+import dev.minutest.rootContext
+import strikt.api.expectCatching
+import strikt.assertions.isA
+import strikt.assertions.isFailure
+
+internal class DeliveryConfigValidatorTests : JUnit5Minutests {
```

Comment:
I wasn't sure if those tests should be at the controller level, or here, as here I can't verify that the config wasn't persisted (...this logic should prevent "bad" configs to even reach `upsert` function)

Reply:
I would leave these tests here, and have separate tests for the controller that mock out the validator, so you can verify that the controller doesn't call `upsert` if validation fails, but the controller tests don't have to deal with all of the different invalid cases.

(I would also remove the "while persisting" language from the contexts below, since you're just testing the behavior of the validator itself in these tests).

Reply:
Maybe just replace "errors while persisting" with "fails validation"?